dex/src/main/java/io/crate/data/SkippingBatchIterator.java

```diff
@@ -22,59 +22,31 @@
 
 package io.crate.data;
 
-import java.util.concurrent.CompletionStage;
+public class SkippingBatchIterator extends ForwardingBatchIterator {
 
-public class SkippingBatchIterator implements BatchIterator {
-
-    private final BatchIterator delegate;
     private final int offset;
-
     private int skipped = 0;
 
-    public static BatchIterator newInstance(BatchIterator delegate, int offset) {
-        return new CloseAssertingBatchIterator(new SkippingBatchIterator(delegate, offset));
-    }
-
-    private SkippingBatchIterator(BatchIterator delegate, int offset) {
-        this.delegate = delegate;
+    public SkippingBatchIterator(BatchIterator delegate, int offset) {
+        super(delegate);
         this.offset = offset;
     }
 
     @Override
     public void moveToStart() {
+        super.moveToStart();
         skipped = 0;
-        delegate.moveToStart();
     }
 
     @Override
     public boolean moveNext() {
         if (skipped < offset) {
             for (; skipped < offset; skipped++) {
-                if (delegate.moveNext() == false) {
+                if (super.moveNext() == false) {
```

Comment:
Maybe have the delegate not passed in the constructor to the super class but have it as a field in SkippingBatchIterator and exposed via a protected delegate() method that the ForwardingBatchIterator will call. This way the ForwardingBatchIterator.moveNext() will call delegate().moveNext() which will make it more obvious that it's not the child class's moveNext that's being called. @dobe What do you think?